python/convert.c

```diff
@@ -0,0 +1,344 @@
+      return PyUpb_GetUint64(obj, &val->uint64_val);
+    case UPB_TYPE_FLOAT:
+      val->float_val = PyFloat_AsDouble(obj);
+      return !PyErr_Occurred();
+    case UPB_TYPE_DOUBLE:
+      val->double_val = PyFloat_AsDouble(obj);
+      return !PyErr_Occurred();
+    case UPB_TYPE_BOOL:
+      val->bool_val = PyLong_AsLong(obj);
+      return !PyErr_Occurred();
+    case UPB_TYPE_BYTES: {
+      char *ptr;
+      Py_ssize_t size;
+      if (PyBytes_AsStringAndSize(obj, &ptr, &size) < 0) return false;
+      *val = PyUpb_MaybeCopyString(ptr, size, arena);
+      return true;
+    }
+    case UPB_TYPE_STRING: {
+      Py_ssize_t size;
+      const char *ptr;
+      PyObject *unicode = NULL;
+      if (PyBytes_Check(obj)) {
+        unicode = obj = PyUnicode_FromEncodedObject(obj, "utf-8", NULL);
+        if (!obj) return false;
+      }
+      ptr = PyUnicode_AsUTF8AndSize(obj, &size);
+      if (PyErr_Occurred()) {
+        Py_XDECREF(unicode);
+        return false;
+      }
+      *val = PyUpb_MaybeCopyString(ptr, size, arena);
+      Py_XDECREF(unicode);
+      return true;
+    }
+    case UPB_TYPE_MESSAGE:
+      PyErr_Format(
+          PyExc_ValueError, "Message objects may not be assigned",
+          upb_fielddef_type(f));
+      return false;
+    default:
+      PyErr_Format(
+          PyExc_SystemError, "Getting a value from a field of unknown type %d",
+          upb_fielddef_type(f));
+      return false;
+  }
+}
+
+bool PyUpb_Message_IsEqual(const upb_msg *msg1, const upb_msg *msg2,
+                           const upb_msgdef *m);
+
+// -----------------------------------------------------------------------------
+// Equal
+// -----------------------------------------------------------------------------
+
+bool PyUpb_ValueEq(upb_msgval val1, upb_msgval val2, const upb_fielddef *f) {
+  switch (upb_fielddef_type(f)) {
+    case UPB_TYPE_BOOL:
+      return val1.bool_val == val2.bool_val;
+    case UPB_TYPE_INT32:
+    case UPB_TYPE_UINT32:
+    case UPB_TYPE_ENUM:
+      return val1.int32_val == val2.int32_val;
+    case UPB_TYPE_INT64:
+    case UPB_TYPE_UINT64:
+      return val1.int64_val == val2.int64_val;
+    case UPB_TYPE_FLOAT:
+      return val1.float_val == val2.float_val;
+    case UPB_TYPE_DOUBLE:
+      return val1.double_val == val2.double_val;
+    case UPB_TYPE_STRING:
+    case UPB_TYPE_BYTES:
+      return val1.str_val.size == val2.str_val.size &&
+          memcmp(val1.str_val.data, val2.str_val.data, val1.str_val.size) == 0;
+    case UPB_TYPE_MESSAGE:
+      return PyUpb_Message_IsEqual(val1.msg_val, val2.msg_val,
+                                   upb_fielddef_msgsubdef(f));
+    default:
+      return false;
+  }
+}
+
+bool PyUpb_Map_IsEqual(const upb_map *map1, const upb_map *map2,
+                       const upb_fielddef *f) {
+  assert(upb_fielddef_ismap(f));
+  size_t size1 = map1 ? upb_map_size(map1) : 0;
+  size_t size2 = map2 ? upb_map_size(map2) : 0;
+
+  if (map1 == map2) return true;
```

Comment:
you can lift the pointer check above the size call

Reply:
Done.